usr/lib/cca_stdll/cca_specific.c

```diff
@@ -1206,6 +1208,288 @@ CK_RV token_specific_rsa_verify(STDLL_TokData_t * tokdata,
     return CKR_OK;
 }
 
+CK_RV token_specific_rsa_pss_sign(STDLL_TokData_t *tokdata,
+                                  SESSION  *sess,
+                                  SIGN_VERIFY_CONTEXT *ctx,
+                                  CK_BYTE *in_data,
+                                  CK_ULONG in_data_len,
+                                  CK_BYTE *out_data,
+                                  CK_ULONG *out_data_len)
+{
+    CK_RSA_PKCS_PSS_PARAMS *pss;
+    long return_code, reason_code, rule_array_count;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0, };
+    long signature_bit_length, message_len;
+    CK_ATTRIBUTE *attr;
+    OBJECT *key_obj = NULL;
+    CK_BYTE *message = NULL;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+
+    rc = object_mgr_find_in_map1(tokdata, ctx->key, &key_obj, READ_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("object_mgr_find_in_map1 failed\n");
+        goto done;
+    }
+
+    /* Find the secure key token */
+    if (!template_attribute_find(key_obj->template, CKA_IBM_OPAQUE, &attr)) {
+        TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+        rc = CKR_TEMPLATE_INCOMPLETE;
+        goto done;
+    }
+
+    pss = (CK_RSA_PKCS_PSS_PARAMS *)ctx->mech.pParameter;
+    if (pss == NULL ||
+        ctx->mech.ulParameterLen != sizeof(CK_RSA_PKCS_PSS_PARAMS)) {
+        TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+        rc = CKR_MECHANISM_PARAM_INVALID;
+        goto done;
+    }
+
+    message_len = 4 + in_data_len;
+    message = malloc(message_len);
+    if (message == NULL) {
+        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+        rc = CKR_HOST_MEMORY;
+        goto done;
+    }
+
+    *((uint32_t *)message) = pss->sLen;
```

Comment:
htonl(pss->sLen) ? I know we are in the CCA token which only runs on z ...

Reply:
Other places of the CCA token also use htox, so I kept it that way. Kind of documents that the CCA structure are always big endian, no matter what platform it is running on.